src/common/iceberg.cpp

```diff
@@ -188,19 +188,32 @@ string GenerateMetaDataUrl(FileSystem &fs, const string &meta_path, string &tabl
 }
 
 
-string IcebergSnapshot::GetMetaDataPath(const string &path, FileSystem &fs, string metadata_compression_codec, string table_version = DEFAULT_VERSION_HINT_FILE, string version_format = DEFAULT_TABLE_VERSION_FORMAT) {
+string IcebergSnapshot::GetMetaDataPath(const string &path, FileSystem &fs, string metadata_compression_codec, string table_version = DEFAULT_TABLE_VERSION, string version_format = DEFAULT_TABLE_VERSION_FORMAT) {
+	string version_hint;
+	string meta_path = fs.JoinPath(path, "metadata");
 	if (StringUtil::EndsWith(path, ".json")) {
+		// We've been given a real metadata path. Nothing else to do.
 		return path;
-	}
-
-	auto meta_path = fs.JoinPath(path, "metadata");
-	string version_hint;
-		if(StringUtil::EndsWith(table_version, ".text")||StringUtil::EndsWith(table_version, ".txt")) {
-		version_hint = GetTableVersion(meta_path, fs, table_version);
-	} else {
+	} else if (!fs.DirectoryExists(meta_path)) {
+		// Make sure we have a metadata directory to look in
+		throw IOException("Cannot open \""+path+"\": Metadata directory does not exist");
+	} else if(StringUtil::EndsWith(table_version, ".text")||StringUtil::EndsWith(table_version, ".txt")) {
+		// We were given a hint filename
+		version_hint = GetTableVersionFromHint(meta_path, fs, table_version);
+		return GenerateMetaDataUrl(fs, meta_path, version_hint, metadata_compression_codec, version_format);
+	} else if(!StringUtil::StartsWith(table_version, "?")) {
```

Comment:
Could you elaborate in a comment what it implies when a table version does not start with `?`

Reply:
If this relates to `DEFAULT_TABLE_VERSION` can we instead turn this into `table_version != DEFAULT_TABLE_VERSION` ?

Reply:
Another artifact of cleaning up my implementation to fix the tests. Originally, I had considered having two different "?" implementations (most recently created vs. lexicographically highest). I will update to the simpler proposal.